### DISCUSSION TITLE
Classification is not robust enough for some prompts.

### DISCUSSION THREAD

**Discussion by zhou13** [olimorris/codecompanion.nvim#541] [Ideas]
It seems that the out of box experience of the inline mode of codecompanion.nvim is not ideal for me: See the following screenshots:

https://github.com/user-attachments/assets/03f30f69-4031-4f04-a890-3411090e5635

1. When I use prompt "add comments", it simply gives "Error", which is really confusing.  After looking into the source code of codecompanion, it seems it is doing classification of the prompt and chatgpt cannot determine that I need "replace" from "add comments". It seems that the prompt is hard coded and cannot be easily changed.
2. After figuring out the problem, I change the prompt to "refactor the code", the performance is still not satifactory. Indentation is wrong and output quality is not good. However, this should be fixable by finetuning the prompts.

In comparison, when I use avante.nvim, I have a good experience with the "add comments" prompt out of the box:

https://github.com/user-attachments/assets/0363d627-0ec1-4a5c-9107-1c7531abaa4f

Overall, I felt like that the current mechanism of adding a classification layer is not robust and overly complex. I am also not sure what is the intention here: If I am using inline mode with visual selection, I always want to replace them and doing a classification is not necessary for me.

**Comment by olimorris** [comment]
Appreciate the feedback and the comprehensive testing. I would welcome any PRs and if you can share the Python file you're working on I'd very much like to test against it.

The primary aim of the inline strategy is to make it more efficient to make simple edits to a buffer. And by simple, I mean the sort of edits that can be easily described in a < 255 chars prompt to an LLM.

Myself and fellow users have reported that inline edits with certain LLMs (often locally hosted ones) can be unreliable. Code is returned in markdown code blocks, or indentations are ignored. However, with newer models this hasn't been so much of an issue. Trying out a simple inline edit in a Python file with Copilot and Claude-3-5-sonnet I can refactor just fine so hence keen to test with your file:

![2024-12-15 23_21_09 - WezTerm](https://github.com/user-attachments/assets/9bae2045-1285-4d02-9022-76c2656cce2c)

What model are you using?

> I am also not sure what is the intent here: If I am using inline mode with visual selection, I always want to replace them and doing a classification is not necessary for me.

The visual selection is about sending context to the LLM. You're selecting what you'd like the LLM to know alongside your prompt. You could of course prefix any prompt with `/buffer` (e.g. `:'<,'>CodeCompanion /buffer refactor this`) to send the whole buffer. I often make a visual selection and ask the LLM to `write a unit test for this`. Of course, _replacing_ this selection wouldn't make any sense hence why the inline prompt has been curated to know that this should result in a new buffer being created to match the filetype of the underlying selection.

> It seems that the prompt is hard coded and cannot be easily changed.

Yep, as you've no doubt seen in the `inline.lua` file, the `Inline:place` method has been written to place output from the LLM in one of [four](https://github.com/olimorris/codecompanion.nvim#pencil2-inline-assistant) ways. It doesn't make sense to put that in the `config.lua` file for users to change. But...I was thinking on my feet back in the early part of this year and no one had done this before so I'm open to suggestions for how this can be improved.

**Reply by zhou13** [reply]
I did more test. I was using openai with default model: `:'<,'>CodeCompanion /buffer add comments` returns "Error" for me. If I change it to claude, then no error is given. But the indentation is still wrong, and it is doing "add" instead of "replace", which mean that the old code was not removed. I uploaded the file used in the video for testing.
[compare-re10k.py.zip](https://github.com/user-attachments/files/18157037/compare-re10k.py.zip)

**Reply by zhou13** [reply]
Regarding the problem of "write a unit test for this.", what do you think to keep them as different commands, or define new parameters (`action='chat'` or `action='replace'`) in the prompt palette so that we don't need extra mappings. It does not save time to use AI to make a decision anyway.

Among the 5 actions:
1. `replace` the current selection
2. `add` after the current cursor position
3. `before` before the current cursor position
4. `new` in a new buffer/file
5. `chat` in a buffer which the user can then interact with

I think `add`, `before`, and `chat` is not necessary. The first two can be replaced with `replace` (AI can add text in the front or at the end), and the `chat` already has a dedicate command.

**Reply by olimorris** [reply]
> what do you think to keep them as (or just add new) different commands, like :CodeCompanionEdit and :CodeCompanionGenerate

I'm unwilling to add new commands to change or even break the workflow for so many users. There's zero to be gained from that. In terms of my personal preference, I do dislike having a plugin with so many different commands especially when most users can accomplish their desired workflow with custom prompts.

For your use case please see the "Specifying a Placement for Inline Prompts" in the [RECIPES.md](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/RECIPES.md) file. As an example, if you had a custom prompt that you call _"Generate"_ you could call it with `:CodeCompanion /generate`. That could even be abbreviated to make it even more terse.

> I think add, before, and chat is not necessary. The first two can be replaced with replace (AI can add text in the front or at the end), and the chat already has a dedicate command.

The first two are necessary if you have custom prompts (#246). The `chat` placement, you're correct, someone could just use `CodeCompanionChat`, it's a convenience factor.

The bigger picture here is that none of the placements *should* be materially affecting the output from the LLM.

**Reply by olimorris** [reply]
Thanks for sharing the python file btw. Can confirm the OpenAI adapter isn't working as I'd expect so I will look into that.

**Comment by olimorris** [comment]
Hey @zhou13.

In case you still use the plugin, I thought I'd let you know that as of `v12.5.0` I've re-built the inline assistant. It's much more performant, robust and with a much clearer ask of the LLM. Basic things like adding comment blocks to functions were a pain in the proverbial.

Really appreciated the time and effort you went to detail your thoughts/findings two months ago.

Oli

**Reply by zhou13** [reply]
Thank you for the update! Will test it again later.
